x-pack/plugins/lens/public/datatable_visualization/components/palette_picker.tsx

```diff
@@ -0,0 +1,447 @@
+
+const idPrefix = htmlIdGenerator()();
+
+type RequiredParamTypes = Required<CustomPaletteParams>;
+
+const DEFAULT_MIN_STOP = 0;
+const DEFAULT_MAX_STOP = 100;
+const MAX_COLOR_STEPS = 12;
+const MIN_COLOR_STEPS = 1;
+const DEFAULT_COLOR_STEPS = 10;
+export const defaultParams: RequiredParamTypes = {
+  name: 'positive',
+  reverse: false,
+  rangeType: 'auto',
+  rangeMin: DEFAULT_MIN_STOP,
+  rangeMax: DEFAULT_MAX_STOP,
+  progression: 'fixed',
+  stops: [],
+  steps: DEFAULT_COLOR_STEPS,
+};
+
+function getPaletteColors(
+  palettes: PaletteRegistry,
+  activePaletteParams?: CustomPaletteParams,
+  // used to customize color resolution
+  { stopFactor, prevPalette }: { stopFactor: number; prevPalette?: string } = {
+    stopFactor: 1,
+  }
+) {
+  // If stops are already declared just return them
+  if (
+    activePaletteParams?.stops != null &&
+    // make sure to regenerate if the user changes number of steps
+    activePaletteParams.stops.length === activePaletteParams.steps
+  ) {
+    return activePaletteParams.stops;
+  }
+  return palettes
+    .get(prevPalette || activePaletteParams?.name || defaultParams.name)
+    .getCategoricalColors(
+      activePaletteParams?.steps || defaultParams.steps,
+      activePaletteParams ?? undefined
+    )
+    .map((color, i) => ({ color, stop: i * stopFactor }));
+}
+
+export function applyPaletteParams(
+  palettes: PaletteRegistry,
+  activePalette?: PaletteOutput<CustomPaletteParams>
+) {
+  if (!activePalette) {
+    return {};
+  }
+  // make a copy of it as they have to be manipulated later on
+  const paletteColorRepresentation = getPaletteColors(
+    palettes,
+    activePalette?.params
+  ).map(({ color, stop }) => ({ color, stop }));
+
+  const paletteDisplayMode = activePalette?.params?.progression
+    ? activePalette?.params?.progression !== 'stepped'
+      ? activePalette?.params?.progression
+      : 'gradient'
+    : 'fixed';
+
+  if (activePalette?.params?.reverse && paletteColorRepresentation) {
+    const stops = paletteColorRepresentation.map(({ stop }) => stop);
+    paletteColorRepresentation.forEach((colorStop, i) => {
+      colorStop.stop = stops[paletteColorRepresentation.length - i - 1];
+    });
+    paletteColorRepresentation.reverse();
+  }
+  return { colorStops: paletteColorRepresentation, mode: paletteDisplayMode };
+}
+
+function mergePaletteParams(
+  activePalette: PaletteOutput<CustomPaletteParams>,
+  newParams: Partial<RequiredParamTypes>
+) {
+  return {
+    ...activePalette,
+    params: {
+      ...activePalette.params,
+      ...newParams,
+    },
+  };
+}
+
+export function CustomizablePalette({
+  palettes,
+  activePalette,
+  setPalette,
+}: {
+  palettes: PaletteRegistry;
+  activePalette: PaletteOutput<CustomPaletteParams>;
+  setPalette: (palette: PaletteOutput<CustomPaletteParams>) => void;
+}) {
+  const [minLocalValue, setMinLocalValue] = useState<string>(
+    '' + (activePalette.params?.rangeMin ?? DEFAULT_MIN_STOP)
```

Comment:
is this a conversion to string? I personally like hacks like that (`''+` or `!!` for boolean conversion) but we're trying to avoid it in Lens. What do you think about just `String(activePalette.params?.rangeMin ?? DEFAULT_MIN_STOP)` here and below in 125 line?